### ISSUE TITLE
Future direction support field

### ISSUE THREAD

**Issue by Eomm** [nodejs/package-maintenance#241]
After the publishing of the draft for `support` field in `package.json` and the feedback process form the community, the ideas to extend moreover that field are:

- reference a file via an URL field
-  a separate `package-support.json` file

reference:
https://github.com/nodejs/package-maintenance/pull/220#issuecomment-516475639
https://github.com/nodejs/package-maintenance/pull/220/files#r310909069

**Comment by ljharb** [comment]
I’d want a separate file - so it can be immutably stored in package registries - but only if the file can be updated without a full publish, and also read without downloading the full tarball.

**Comment by Eomm** [comment]
The problem is that npm let you download only the package.json (in two sizes, one minimal with less information, and one with all the metadata).
If you want something else you need to download the tarball

**Comment by ljharb** [comment]
Right - that’s the reason why, until npm adds a feature or two, package.json is the only place it should go.

**Comment by kemitchell** [comment]
@ljharb, thanks for pointing me here from https://github.com/npm/cli/pull/187. I hope folks don't mind a few stray thoughts and up-to-speed questions.

I think it's safe to say that a great many package consumers rely heavily on URL-based pointers from `package.json` files already. Specifically, once we know that `repository` has a GitHub or BitBucket URL, we know that's our best bet for up-to-date information on bugs, changelog, patch submission, and consumer-to-producer comms more generally. When we open the code host page, we can look to `README.md`---the latest on `master`, not the one that came from the registry---for CI badges, documentation site links, chat channel buttons, and so on. There's just no guarantee that yesterday's hot new channel will make it to the next version. We've all seen patch releases replacing Slack channels with Gitter, or Gitter with Slack, or Slack with IRC or just issues again. A lot of us have tried a few different sustainability-as-a-service platforms, and forgotten to update some of our packages.

If you're trying to do automated dependency assessment, perhaps by modeling risk on the availability or throughput of support channels, I don't think the pain point is lack of information. Rather, it's that some dev has to sift through all the repo webpages, click all the right links, and tally up all the actually relevant URLs.  Much the same story if you're trying to take some kind of action---send thank-you e-mails, send donations, ask for support contracts---across many packages. Some of what `PACKAGE-SUPPORT` seems to be gunning for is structured data for facilities that haven't been well structured before, like support channels. But big chunks of it, specifically `backing`, basically map packages to other URLs.

How do we anticipate people using the data, wherever developers end up storing it? Is it safe to assume that requests for this data (`npm support`?) will be far less frequent than requests to install the packages (`npm install`)? Is it safe to assume that people will get to specific package versions through their `node_modules` directories, or will they come to packages _as packages_, with an assumption they'll keep up-to-date with the latest release?

I suspect the back-and-forth on where to put this data, no matter how it's encoded, reflects unexplored differences in how you imagine the data being consumed. But that's a guess. I'm a recent interloper here. ;-D

**Comment by wesleytodd** [comment]
> I hope folks don't mind a few stray thoughts and up-to-speed questions.

The more input and questions the better!

> I think it's safe to say that a great many package consumers rely heavily on URL-based pointers from package.json files already

You make a good point, but I would also like to add the counter argument that many package links are broken.  I often click github links which have are a 404, I am sure the other cases are similar.  I don't say this as a blocker to the idea, just that it is not a perfect solution.

> Some of what PACKAGE-SUPPORT seems to be gunning for is structured data for facilities that haven't been well structured before

Yes this is IMO one of the main goals of this proposal.  If we have structure we can write reliable tooling, educate the community on it in a reasonable way, and help a broad range of people instead of each group going in their own direction.

> But big chunks of it, specifically backing, basically map packages to other URLs.

"Basically" I agree, but the devil is in the details.  I think the structure we have defined gives more than *just* a list of links.  The difference between "I want to donate to projects I use" and "I want to use projects which offer paid support plans" is quite different.  Just a list of links does not help answer this question.

For example, I think a company might ask the second question, and IMO getting companies to "pony up" is one of the best ways to raise meaningful financial support for projects.  Having this explicitly in the spec is important for this goal IMO.

> Is it safe to assume that requests for this data (npm support?) will be far less frequent than requests to install the packages (npm install)?

Well, if we do it well I hope that we can do something like `funding` was doing where it only showed once and is a relatively clear and clean UX.  So if we do that, then we can run `npm support` on all installs, and it would just show the new or important information to users.  Very similar to `npm audit`'s behavior.

> Is it safe to assume that people will get to specific package versions through their node_modules directories, or will they come to packages as packages, with an assumption they'll keep up-to-date with the latest release?

I can see running `npm support package-name` and have it look locally then remotely.  I can also see this interaction also being something that always does a network request to get the up-to-date file.  Maybe doing `npm support --local` so that you can run w/o network.

I just think we can evolve the use case over time as we see clear reason. I think the important thing is making sure we don't make any of these impossible in the future.  So to me, allowing url's which are not resolved into `node_modules` excludes the use case of `--local`, which to me means we should not allow it in the spec for now until we can understand more of the use cases.

**Comment by Eomm** [comment]
For tracking: the discussion in https://github.com/npm/cli/pull/246 is continuing with a wide feedback from Isaacs

**Comment by wesleytodd** [comment]
Also, if anyone here wants to chime in there and provide different viewpoints than the one I am expressing, it would be helpful :)

**Comment by mhdawson** [comment]
The --local option is interesting, although my first thought is that if I was using the tool to verify support levels for the packages I'm going to deploy I would want to ensure I have the latest info.  Having said that the info can change immediately after I check it so not having fully up to date info must be ok in some situations as well.

**Comment by wesleytodd** [comment]
> not having fully up to date info must be ok in some situations as well.

I am not sure we can get around this, we are just in a world of imperfect information.  But anything is better than what we have today.  With tooling at least we can monitor and check for changes over time.

**Comment by mhdawson** [comment]
> I am not sure we can get around this, we are just in a world of imperfect information. But anything is better than what we have today. With tooling at least we can monitor and check for changes over time.

+1

**Comment by Eomm** [comment]
Adding this to the agenda since the feedback from npm has a great impact on how to continue in this issue, in order to don't create confusion in the community of course

Ref: https://github.com/nodejs/package-maintenance/issues/241#issuecomment-529088119

**Comment by Emuentes** [comment]
In our meeting today, we discussed the possibility of leaving BOTH options open, allow people to choose whether they want to drop a link or a valid JSON property within the package.json

Is there a reason we absolutely MUST pick one or the other before sharing our recommendation for the general structure of the support property?

**Comment by ljharb** [comment]
Yes, I remain opposed to allowing anything that's not inline, until such time as npm allows updating of metadata without a version bump.

**Comment by lholmquist** [comment]
> until such time as npm allows updating of metadata without a version bump.
This is probably a discussion for another thread,  but i feel like if that happens, then that is breaking the contract of that particular version, even if it is metadata and doesn't affect the functionality of what the code does.   Sorry for the noise, but just wanted to get this thought out there

**Comment by ljharb** [comment]
This would have to be metadata separate from the version tarball.

**Comment by ghinks** [comment]
# Financial Technology (FinTech)

On the lines of the

- reference a file via an URL field
-  a separate `package-support.json` file

there was a discussion that I participated in at the package maintenance meeting of the 24th Sept which I was asked to document. I am going to call this the **behind the firewall** issue.

Most FinTech financials that operate within our community do so in a secure environment. It is a **mistake** to think that the same **firewall** issues that exist at technology companies apply to  **FinTech**.  They are different.

Within the financial sector all environments not just **production** are secure. This would mean that using a **URL** may exclude many targeted users from actually seeing the support levels.

Each and every **FinTech** organization would have its own rules and levels of comfort but in the spirit of wanting to bring some of the larger organizations into the family circle to participate in our community a local file provides local parsing support without the need to reach outside of the organization.

I am however noting that as this is a new item that we are defining I do not believe the use of the URL would be an initial issue to FinTech. Only once widely adopted may it prove to be so.

**Comment by kemitchell** [comment]
@ghinks what you say is true of large financial institutions---and not just financial technology companies---in my experience. But I don't think it's on the rest of the economy to design around their choice to secede from the Internet.

**Comment by ghinks** [comment]
Indeed @kemitchell it is not and we should not pander to any particular sector of our users to the detriment of the wider community. It is however just another useful piece of the moziac :)

**Comment by mhdawson** [comment]
One more idea we had today when we were discussing the blog post to get more feedback:

* Ask for both inline and url (ie URL which is link to latest version of the data as well as json for the current version) in the package.json  (this is different suggestions which were OR, it is AND)
* This does require an file external to npm to host the data, but github should be a good option
* At some point in the future we'd hope that the npm client could update the package.json on publish with the latest from the link automatically.
* This ensures that the data is available both online and offline without adding to the load on npm in terms of fetching data.


@ljharb what do you think?

**Comment by ljharb** [comment]
Duplicating it means you have two sources of truth; i'm not sure why there's any value in that.

I think that npm would actively *not* want to update the package.json - rather, some other file/metadata.

**Comment by wesleytodd** [comment]
@ljharb Do you think you would be available to come to the next [npm rfc meeting](https://github.com/npm/rfcs/issues/52) when they schedule it?

I was doing my best to represent all the facets of the conversation we have had in the WG to npm, but I think with you being the champion of this particular part it would be better if you and the folks at npm could talk directly.

Or maybe seeing if @darcyclarke (and @isaacs maybe?) can attend our next regular meeting?  Either one would be good to help us all get on the same page.

**Comment by ljharb** [comment]
@wesleytodd I'd love to, but they seem to always be at 1pm ET on Wednesdays, which is a time I can't make.

**Comment by mhdawson** [comment]
@ljharb the suggestion was in the spirit of a `compromise` there are many people who feel a url is better, more up to date.  In the suggestion it is only a duplicate if it has been update and is considered the "source of truth".  At the same time it tries to address your concerns that the data be available off line and from the npm registry.   The suggestion about npm updating  (in the client as it was being published) was so that what was published would be the most recent data.

**Comment by ljharb** [comment]
If there's a URL, then the inline data won't likely be kept up to date, and then my concern is simply not addressed. The only way it will always be up to date is if it's *forced* to be part of the package.

**Comment by mhdawson** [comment]
The suggestion to have it be auto updated on publish from the url was intended to ensure it was as up to date as possible.

**Comment by ljharb** [comment]
If that’s something npm can support, and it’d error the publish if the data was unavailable or invalid, and it was ergonomic to provide both or one or the other, and if a mismatch between the url and the online data errored too, then that does seem like a workable compromise.

**Comment by mhdawson** [comment]
@ljharb you might know better than me but it should be something that somebody outside of the npm could PR into the client?

**Comment by ljharb** [comment]
Surely anyone could write the PR, but the key question is, would they accept it :-)

**Comment by mhdawson** [comment]
@darcyclarke any comments on the suggestion in: https://github.com/nodejs/package-maintenance/issues/241#issuecomment-540748913

**Comment by ghinks** [comment]
@Eomm when you said that there are 2 types of package.json that can be downloaded on August 28th one with meta data how are the mechanisms different to fetch them, I ask as I thought plan old curl https://registry.npmjs.org/expressjs was the only option. ( just part of my background on the blog for support :) )

**Comment by Eomm** [comment]
I was referring to [this doc](https://github.com/npm/registry/blob/master/docs/responses/package-metadata.md).

For what I saw, the full version is the double time the lighter's version in the worst case.

**Comment by ghinks** [comment]
I made a few updates to the [google draft document as we suggested](https://docs.google.com/document/d/1-2iLaVEQCwQeCmtQ1fvgVDLOPwUZpUUY-3cfQoMUTvo/edit?usp=sharing). 

After some time for reflection, which is always good, I am hoping we can reach a compromise. As I do not believe either method could be fully satisfied currently without getting the registry involved.

**Comment by darcyclarke** [comment]
> @darcyclarke any comments on the suggestion in: [#241 (comment)](https://github.com/nodejs/package-maintenance/issues/241#issuecomment-540748913)

I'll try

> * Ask for both inline and url (ie URL which is link to latest version of the data as well as json for the current version) in the package.json (this is different suggestions which were OR, it is AND)

I'm not sure I understand the benefit of doubling up on where this information lives, even in the interim.

> * At some point in the future we'd hope that the npm client could update the package.json on publish with the latest from the link automatically.

I'm a bit concerned with having `npm` agree to, at some point in the future, pull in any data automatically for a publisher; I feel like that strategy would be prone to conflicts and cause us a bit of pain from the tooling/support side. 

> * This ensures that the data is available both online and offline without adding to the load on npm in terms of fetching data.

I'm not sure if package maintainers would want to keep two copies of the same information up-to-date even in the interim (not sure if that's what's being suggested here and, again, that's just a subjective opinion).

---

On a separate yet related note, I've [filed an `npm` RFC](https://github.com/npm/rfcs/pull/54) for a much smaller chunk of work to support a `funding` field in `package.json`. As I've noted in a few of the various calls we've had, I believe that the current scope of the `support` draft is trying to achieve a bit too much at one time. 

I believe that focusing on the monetary opportunities, first, for maintainers helps move the needle forward and aligns with `npm`'s [commitment, we made to the community, at the end of August](https://blog.npmjs.org/post/187382017885/supporting-open-source-maintainers). I've taken some of the concepts outlined for the `backing` field, specifically, and scoped what the tooling would look like on our end.

I'd love to discuss this more in tomorrow's meeting with the hope that other active/vocal community members can join or provide their own feedback (ie. @ljharb @feross @arcanis @zkochan). Please let me know if there's anything I need to do to specifically add this as an agenda item. <3

**Comment by ljharb** [comment]
We had a great discussion in our meeting today.


Following that, @mhdawson and I had a productive chat, and here's what we'd like to propose: a package.json field called `support` which is one of: 

* `true`: This will imply your package.json's `repository` URL (including `directory`), with a default path value of `./package-support.json`
* a string: a relative path, starting with `./`, that points to a file within the repository specified in the existing repository field
* object, with `repository` key matching the schema for the existing repository field, and optional `path` field that if provided, must be a relative path, starting with `./`, within the specified repository.

The content of the file will match the schema we’ve already discussed.

Here’s the use cases we envision:

1. common case: `"support": true` with an associated `package-support.json` file in your repo at the top level.
1. much less common case: an alternate file path within the package’s repo, such as `"support": "./path/to/support.json"`
1. much much rarer case: sharing the same support data across multiple repos. Examples might be
  - `”support”: { “repository”: { “type”: “git”, “url”: “node/.github” }, “path”: “support.json” }`
  - `”support”: { “repository”: { “type”: “git”, “url”: “node/.github” } }`
  - `”support”: { “repository”: { “type”: “git”, “url”: “node/.github”, “directory”: “packages” } }`

These will resolve to the location of a JSON file (called `package-support.json` by default) that matches the schema discussed above.

Our recommendation will be to use our tooling to manage `package.`json, and to be sure to include `package-support.json` (or your alternative filename) in your `files` array, or, be sure that `.npmignore` does not exclude it.

The effect of this is that, by default, each publish will contain a snapshot of the support data, but the true source of truth is always a file within the package’s repository (or within an external repository). This supports offline and “no external internet” use cases as best as can be achieved.

By using a repository, we are using a known set of protocols (git, svn, etc), and creating a pit of success for folks to both store their support data in version control, and also include a snapshot of it in their published packages, with no additional effort.

Our tooling will not allow accessing any arbitrary repository URL by default - to avoid privacy concerns, we will survey npm packages for the most common repository hosts (presumably, github, bitbucket, gitlab, etc), and allow-list these by default. Users of our command line tool will be able to override the allowlist, or bypass it, as desired.

In the future, if npm sees value in this field, they could, for example, add validation into `npm publish`, surface this information on npmjs.com, and/or inline the support data into the packument so that `npm view $package support` or similar produces the data.

**Comment by arcanis** [comment]
> I'd love to discuss this more in tomorrow's meeting with the hope that other active/vocal community members can join or provide their own feedback (ie. @ljharb @feross @arcanis @zkochan).

Thanks for the ping! I'm currently in vacations so attending is difficult, but I appreciate being at least aware of the discussion 👐

The proposal described by @ljharb would solve the main issue I had with this proposal. 👍

My only remark is that the `repository` field already has a way to specify [a path within a repository](https://github.com/greysteil/rfcs/blob/9bc5c6ab0f14fc2c2455484b49415f1412323598/accepted/0010-monorepo-subdirectory-declaration.md). It currently doesn't have any effect, but we've been open to leverage it to connect a package to its real location (for example to print the readme / changelog). In this context, I think it would be a good idea to specify where the path is relative from when a `directory` field is present (and I'd suggest it to be the package directory rather than the root).

**Comment by ljharb** [comment]
@arcanis the intention is to respect the directory field (that the npm website already respects afaik); my comment includes an example of doing so.

I’m afraid i can’t attend the npm meeting; so far it’s always been Wednesday at 9am PT which i have a standing conflict for.

**Comment by lholmquist** [comment]
@ljharb that outline looks really great.  

I thought i had while reading is would it be beneficial to try and get the "common case" creation supported in `npm init`?  probably another issue to discuss

**Comment by ljharb** [comment]
Absolutely it would :-) but that’s something we can do iteratively, after we’ve demonstrated some community adoption of the support field via our own tooling.

**Comment by mhdawson** [comment]
@nodejs/package-maintenance can you chime in on the latest proposal in: https://github.com/nodejs/package-maintenance/issues/241#issuecomment-545178197.  If we can get consensus on this we can move on to the next phase of getting broader input.

**Comment by kemitchell** [comment]
Folks: Just FYI, I have to unsubscribe here. I'm slammed now, but finally getting some desperately needed time off next week.

Please feel free to @-mention me if there's some specific place where I can help. But please understand that I won't be responding for the next week or so.

**Comment by Eomm** [comment]
> This supports offline and “no external internet” use cases as best as can be achieved.

Is this assuming the "no external internet" has access to the most common repository hosts right?
I just want to be sure to have understood all the cases.

**Comment by ljharb** [comment]
No, in those scenarios, they would be relying on the contents of the tarball.

**Comment by Eomm** [comment]
So should the 3rd use case ("sharing the same support data across multiple repos") download the file on installation?

I don't understand when the fetch happen in that scenario

**Comment by ljharb** [comment]
Yes, ideally - but i don’t think that will be a common use case, and our tooling could still download it prepublish for those making that extra effort.

**Comment by mhdawson** [comment]
Next steps:
* PR the latest suggestion from @ljharb and @mhdawson into the doc. (Michael)
* Update blog post
* Start brainstorming/working on tooling to support.
* Promote

**Comment by wesleytodd** [comment]
> Start brainstorming/working on tooling to support.

This was the compliant JSON schema for the previous version.  Maybe update that with the changes?

**Comment by Eomm** [comment]
> This was the compliant JSON schema for the previous version. Maybe update that with the changes?

Yes, I think the direction more stable now and we could improve the [pkgjs/support](https://github.com/pkgjs/support) package: do you agree to open some issue there to split the job?
